### DISCUSSION TITLE
Dynanmic schemas

### DISCUSSION THREAD

**Discussion by lveillard** [colinhacks/zod#1099] [Q&A]
Hello all! 
What would be the best way to create schemas on runtime? 

Let's say the user can add fields, edit the types and add other validations on the fly (or maybe store them on DB)

As an example let's say the user drag and drops a number input and adds a validation max=20, i would need to:
1) Add `newNumber` to the schema object
2) chain `.number().max(20)` to it
3) Hardcore mode: Add some custom validation that's linked to another field thats also dynamicm for example min = 0 if the user adds a boolean field `onlyPositive`, and set it to `true`
4) Apply the changes.

Alternatively these validation rules could arrive dynamically from DB

Any ideas? 
Thanks!

**Comment by devjabeer** [comment]
Hi,
Are you able to solve this issue? I'm looking for similar implementation. Thanks

**Reply by lveillard** [reply]
Nope sorry! I will probably create a custom validation engine adapted to my particular need

**Comment by mcky** [comment]
There's no reason you can't programatically generate a schema - zod won't be able to statically infer the return types but it can still parse and validate

The below shows how you can dynamically create a schema and also add dynamic constraints.

It'd currently blow up if you added something like `{ email: true }` since there's no `.number().email()` - but you could certainly parse `formData` itself with a zod schema!

```ts
import { z } from "zod";
import assert from "assert";

type FieldBase = { name: string };
type NumberField = FieldBase & {
  type: "number";
  validation?: { max?: number; min?: number };
};
type BooleanField = FieldBase & { type: "boolean"; validation?: {} };
type FieldDef = NumberField | BooleanField;

const formData: { fields: FieldDef[] } = {
  fields: [
    { type: "boolean", name: "onlyPositive" },
    { type: "number", name: "newNumber", validation: { max: 20, min: 10 } },
  ],
};

const chainValidationParams = (schema, validation: Record<string, any>) => {
  return Object.entries(validation).reduce((acc, [k, v]) => {
    return acc[k](v);
  }, schema);
};

const fieldToSchema = (field: FieldDef) => {
  let schema;
  if (field.type === "number") {
    schema = z.number();
  } else if (field.type === "boolean") {
    schema = z.boolean();
  }
  
  return chainValidationParams(schema, field.validation || {});
};

const schemaKeys = formData.fields.reduce((acc, field) => {
  return {
    ...acc,
    [field.name]: fieldToSchema(field),
  };
}, {});

const output = z.object(schemaKeys).parse({
  newNumber: 20,
  onlyPositive: true,
});
 
assert.deepEqual(output, {newNumber: 20, onlyPositive: true})
```


I couldn't really make sense of the `onlyPositive` requirement. It sounds like your use case is a form builder and `onlyPositive` sounds like a field definition, which then wouldn't have a value?
